source/adapters/level_zero/event.cpp

```diff
@@ -935,7 +951,12 @@ ur_result_t CleanupCompletedEvent(ur_event_handle_t Event, bool QueueLocked,
   std::list<ur_event_handle_t> EventsToBeReleased;
   ur_queue_handle_t AssociatedQueue = nullptr;
   {
-    std::scoped_lock<ur_shared_mutex> EventLock(Event->Mutex);
+    // If the Event is already locked, then continue with the cleanup, otherwise
```

Comment:
This looks like a workaround. What is causing the recursive locking?

Reply:
Due to the faster completion, if one is using the host proxy event ie the event that had device scope was waited on using urEventWait, then we call GetOrCreateHostProxyEvent. This grabs the lock for the event and tries to create an event that waits on that event to complete with host scope. However, if that event is completed already, then we will call "cleanup event" which means we will take the lock again. This will be a recursive lock and fail. To fix this, we now do a try lock and also check the "bool" that is set and unset while under the lock to see if this is needed. This is because the check for lock ownership fails so another method was needed.